### DISCUSSION TITLE
LineBotが反映されない

### DISCUSSION THREAD

**Discussion by De-chi** [okuokuch/pokemon_sv_vs_support#116] [Q&A] [answered]
Googleドキュメントの記載通りにスプレッドシートのコピーをし、LINEBotの導入をしたが、Lineビジネスで作成したBotに文字を入れてもBotが反応せず、スプレッドシートへの書き込みも行われない。
これに関して考えられる原因はありますか？
とりあえずドキュメント記載内容を再度確認し、その通りにやったつもりではありますが…

**Comment by okuokuch** [comment]
@De-chi 
考えられる原因はたくさんあるのですが、以下確認お願いできますでしょうか？
- Google スプレッドシートとの連携の手順13の検証ボタンを押した際に、成功と表示されたか？

上記が表示されている場合、Apps Scriptsに記述するアクセストークンか、スプレッドシートIDの記述が間違っている可能性が高いです。これらが正しいかもう一度確認していただけますでしょうか？

**Reply by De-chi** [reply]
@okuokuch 
<img width="855" alt="image" src="https://user-images.githubusercontent.com/125797028/219947081-d31bf7f5-271a-48d6-b389-92d7c6fe97a3.png">
写真反映されるかわかりませんが「成功」は出ています。

アクセストークンはLineDevelopers内設定「メッセージングAPI」の下部に記載のチャネル アクセストークンのURLをすべてコピペ、スプレッドシートIDは「log」シートのURLで、d/より後～/editの手前部分のコピペでお間違いないでしょうか？

**Reply by De-chi** [reply]
@okuokuch もし差し支えなければ私のアクセストークンのURLの部分とスプレッドシートIDのURLの記載部分をスクリーンショットで撮影してこちらに貼り付けようと思います。

**Comment by okuokuch** [comment]
@De-chi 
画像ありがとうございます。デプロイの連携はできていそうですね。

> もし差し支えなければ私のアクセストークンのURLの部分とスプレッドシートIDのURLの記載部分をスクリーンショットで撮影してこちらに貼り付けようと思います。

セキュリティ的にあまりオープンにすべき情報ではないので、一旦共有はなしでお願いします！

以下確認お願いします。
図のように、マクロ.gsの横に黄色い◯がついていますでしょうか？ついている場合ファイルの保存がされておらず、保存前のものをデプロイしてしまっています。
コントロールボタン+Sボタンを押すか、実行ボタンの左にあるプロジェクトの保存ボタンでマクロ.gsの保存を行ってから、再度手順書の8以降の操作をしていただいてよろしいでしょうか？
![image](https://user-images.githubusercontent.com/94620772/219948076-4c1760d0-5713-40e1-95e8-23d9fe8f52b7.png)

もし、上記で黄色のボタンの表示がなされてなければ、問題点見つけ出すために追加で指示しようと思っています。

**Comment by De-chi** [comment]
@okuokuch 
ありがとうございます。
<img width="258" alt="image" src="https://user-images.githubusercontent.com/125797028/219949764-0d427388-812d-4d8a-b29f-751058f52b5c.png">
もう一度やってみましたが、添付画像のようになります。

**Comment by okuokuch** [comment]
@De-chi 
確認ありがとうございます。
では、原因切り分けしていきたいと思います。まず、LINE Bot自体が動いているかの確認です。
1. マクロ.gsの13~65行目の`function doPost(e) {`から始まって`}`で終わる部分を以下に上書きする。
```
function doPost(e) {
    var json = e.postData.contents;
    var replyToken = JSON.parse(json).events[0].replyToken;
    var message = JSON.parse(json).events[0].message.text;

    UrlFetchApp.fetch(
        URL, {
        'headers': {
            'Content-Type': 'application/json; charset=UTF-8',
            'Authorization': 'Bearer ' + ACCESS_TOKEN,
        },
        'method': 'post',
        'payload': JSON.stringify({
            'replyToken': replyToken,
            'messages': [{
                'type': 'text',
                'text': message,
            }],
        }),
    }
    );

    return ContentService.createTextOutput(JSON.stringify({ 'content': 'post ok' })).setMimeType(ContentService.MimeType.JSON);
}
```
2. 手順書の8~16を実行する。(マクロ.gs保存して、新しいデプロイ、Webhook設定の流れです。)

上記設定が正しくできていた場合、LINE Botに送った文章がそのまま返ってきます。

**Reply by De-chi** [reply]
@okuokuch 
こちら行数かなり削減されると思いますが添付画像の文章をそのままコピペでよろしいのでしょうか？

**Reply by okuokuch** [reply]
@De-chi 
そのままコピペで大丈夫です！

補足ですが
もとは、LINE Botへのメッセージをスプレッドシートに反映させて、その旨をLINEのメッセージとして送信するという処理です。
上記メッセージの文章は、送られてきたものをそのまま返すという処理です。
ほとんど何も処理していないので、行数が減っています。

**Reply by okuokuch** [reply]
上記を行う理由について補足です。トラブルの原因が2つあると思っています。
- アクセストークン関連
- スプレッドシート関連

上記メッセージで示した処理だと、スプレッドシートは全く関係ないので、アクセストークン回り関連のトラブルであることがわかります。

**Reply by De-chi** [reply]
@okuokuch なるほど、、、
ですが、こちらを行った場合スプレッドシートIDを入れる行がなくなってしまいます。
どうやら私が見ているものだと17行目にURLを入れるようになっていて、、、
問題ないのでしょうか？

**Reply by okuokuch** [reply]
@De-chi 
はい問題ありません！
というのも、スプレッドシートの影響を省きたいので、17行目を消してます。
スプレッドシートと全く関係ないところでのエラーかどうかを判定したいということです。

**Reply by De-chi** [reply]
@okuokuch 
なるほど、、、
取り急ぎコピペしてみましたが特に変化はありません。

**Reply by okuokuch** [reply]
@De-chi 
なるほど。
となると、アクセストークンが間違っている可能性がありますね。いったん個別でDMさせていただきます。

**Comment by okuokuch** [comment, marked as answer]
@Syumiru 
こちらの件、アクセストークンが間違っていたと言うことがわかりました。
再度、最初から設定してもらう予定なのですが、それでも動かない場合に、しゅみるさん目線で間違えそうなポイントあれば、補足お願いしても良いでしょうか？

@De-chi 
再度、スプレッドシートのコピーから設定してもらう事可能でしょうか？

**Reply by Syumiru** [reply]
承知しました。
